chain-signatures/contract/tests/tests.rs

```diff
@@ -371,3 +396,144 @@ async fn test_contract_initialization() -> anyhow::Result<()> {
+    test_propose_update_contract(&contract, &accounts).await;
+}
+
+async fn test_propose_update_config(contract: &Contract, accounts: &[Account]) {
+    // contract should not be able to propose updates unless its apart of the participant/voter set.
+    let execution = contract
+        .call("propose_update")
+        .args_json(serde_json::json!({
+            "contract": vec![1, 2, 3],
+        }))
+        .transact()
+        .await
+        .unwrap();
+    dbg!(&execution);
+    assert!(execution.is_failure());
+
+    // have each participant propose a new update:
+    let new_config = Config {
+        triple_timeout: min_to_ms(10),
+        ..Default::default()
+    };
+    let mut proposals = Vec::with_capacity(accounts.len());
+    for account in accounts {
+        let execution = account
+            .call(contract.id(), "propose_update")
+            .args_json(serde_json::json!({
+                "config": &new_config,
+            }))
+            .transact()
+            .await
+            .unwrap();
+        assert!(execution.is_success());
+        dbg!(execution.logs());
+        let proposal_id: UpdateId = execution.json().unwrap();
+        dbg!(&proposal_id);
+        proposals.push(proposal_id);
+    }
+
+    let old_config: Config = contract.view("config").await.unwrap().json().unwrap();
+    let state: mpc_contract::ProtocolContractState =
+        contract.view("state").await.unwrap().json().unwrap();
+
+    // check that each participant can vote on a singular proposal and have it reflect changes:
+    let first_proposal = &proposals[0];
+    for (i, voter) in accounts.iter().enumerate() {
+        dbg!(voter.id());
+        let execution = voter
+            .call(contract.id(), "vote_update")
+            .args_json(serde_json::json!({
+                "id": first_proposal,
+            }))
+            .transact()
+            .await
+            .unwrap();
+
+        // NOTE: since 2 out of 3 participants are required to pass a proposal, having the third one also
+        // vote should fail.
+        if i < 2 {
+            assert!(
+                execution.is_success(),
+                "execution should have succeeded: {state:#?}\n{execution:#?}"
+            );
+        } else {
+            assert!(
+                execution.is_failure(),
+                "execution should have failed: {state:#?}\n{execution:#?}"
+            );
+        }
+    }
+    // check that the proposal executed since the threshold got changed.
+    let config: Config = contract.view("config").await.unwrap().json().unwrap();
+    assert_ne!(config, old_config);
+    assert_eq!(config, new_config);
+}
+
+async fn test_propose_update_contract(contract: &Contract, accounts: &[Account]) {
+    let state: mpc_contract::ProtocolContractState =
+        contract.view("state").await.unwrap().json().unwrap();
+
+    // Let's propose a contract update instead now.
+    let new_wasm = std::fs::read(CONTRACT_FILE_PATH).unwrap();
```

Comment:
Technically, it is the same contract.
Not blocking, but we will need to have a test, that the contract was updated. More than that, it would be great to have a test, that this new contract is also upgradable :)

Reply:
I'll add a status_msg contract to show that it does indeed change if we send in a different contract. I'm trying to figure out a way to safeguard against a case where we deploy unwanted contract. In that case, I think we should just validate the core interfaces to see that they still exist like:
- state
- sign
- respond

Reply:
+1 to validating interface. We should probably also check that contract is still upgradable, otherwise it's locked

Reply:
the contract will always be upgradeable since there's no stopping the contract itself from calling deploy on itself even if the contract is locked

Reply:
yes, but if we break the function that is triggering this logic - we will loose the contract